server/src/hover.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) ZeroC, Inc.
+
+use slicec::{
+    compilation_state::CompilationState,
+    grammar::{
+        Class, CustomType, Enum, Enumerator, Exception, Field, Interface, Module, Operation,
+        Parameter, Primitive, Struct, Symbol, TypeAlias, TypeRef, TypeRefDefinition, Types,
+    },
+    slice_file::{Location, SliceFile},
+    visitor::Visitor,
+};
+use tower_lsp::lsp_types::{Position, Url};
+
+pub fn get_hover_info(state: &CompilationState, uri: Url, position: Position) -> Option<String> {
+    let file_path = uri
+        .to_file_path()
+        .ok()?
+        .to_path_buf()
+        .as_path()
+        .to_str()?
+        .to_owned();
+
+    let file =
+        state
+            .files
+            .get(&file_path)
+            .expect(&format!("{:?} {:?}", file_path, state.files.keys()));
+
+    // Convert position to row and column 1 based
+    let col = (position.character + 1) as usize;
+    let row = (position.line + 1) as usize;
+
+    let mut visitor = HoverVisitor::new(slicec::slice_file::Location { row, col });
+    file.visit_with(&mut visitor);
+
+    visitor.found_message
+}
+
+struct HoverVisitor {
+    pub search_location: Location,
+    pub found_message: Option<String>,
+}
+
+impl HoverVisitor {
```

Comment:
I'm not convinced that using the visitor pattern is the right call here. We have to keep visiting the AST elements after we find the definition and check that there is nothing to do. Can we instead look into improving the AST to provide the definition within a location?

Reply:
Yeah, I think that's something we should improve in the future.

Reply:
> Can we instead look into improving the AST to provide the definition within a location?

Yes, but it probably wouldn't be any faster. The AST is just a `Vec<Node>`, it doesn't store which nodes came from which files. So you'd spend alot of time checking nodes from a totally unrelated file.

In the future, we could replace this visitor with a lower-level check:
```
for definition in &file.contents() {
    // Check the definition's span and return when we find one.
}
```